### PR TITLE
feat: implement metadata extraction

### PR DESCRIPTION
**Pull request by CristiCanizales**
### What does this PR do?
- Calls LLM to generate document
- Calls Apex LSP to validate eligibility

### What issues does this PR fix or reference?
@W-17236645@

### CODE REVIEW

File: packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts [hunk 1]
```diff
@@ -120,7 +116,7 @@ export class MetadataOrchestrator {
       let cursorPosition;
       if (isMethodSelected) {
         const editor = vscode.window.activeTextEditor;
-        if (editor && editor.document.fileName.endsWith('cls')) {
+        if (editor && editor.document.fileName.endsWith('.cls')) {
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts:119]:
``` suggestion
        if (editor?.document.fileName.endsWith('.cls')) {
```

File: packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts [hunk 2]
```diff
@@ -177,18 +173,22 @@ export class MetadataOrchestrator {
...
-      result = await apiClient.naturalLanguageQuery({
-        prefix: '',
-        suffix: '',
-        input,
-        commandSource: CommandSource.NLtoCodeGen,
-        promptId: 'generateOpenAPIv3Specifications'
-      });
-      documentContents = result[0].completion;
-      if (documentContents.includes('try again')) throw new Error(documentContents);
+      while (!documentContents.startsWith('yaml') && tries < 10) {
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts:180]: Why the retries?

Comment (CristiCanizales) [reply packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts:180]: It's just a temporary workaround to not have to run the command multiple times until the doc is finally generated

File: packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts [hunk 3]
```diff
@@ -42,182 +44,26 @@ export interface Parameter {
...
-          }
-          isAuraEnabled = false;
-        }
-
-        // Reset for the next method
-        currentMethodSignature = '';
+    if (!isEligibleResponses[0].isEligible) {
+      if (isMethodSelected) {
+        const name = isEligibleResponses?.[0]?.symbols?.[0]?.docSymbol.name;
+        throw new Error(nls.localize('not_aura_enabled', name));
```
Comment (mingxuanzhangsfdx) [inline packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts:62]: could the error be due to wrong access modifiers or so?

Comment (CristiCanizales) [reply packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts:62]: I don't know, I think there's no way to tell with only `.isEligible`. We could modify the message to include that it's wither not auraenabled or private, or something like that

Comment (mingxuanzhangsfdx) [reply packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts:62]: Yeah we could provide a more general error message with help of Sonal.

File: packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts [hunk 4]
```diff
@@ -88,34 +91,17 @@ export class ApexActionController {
...
    * @returns The OpenAPI document as a string.
    */
-  private generateOpenAPIDocument = (metadata: MethodMetadata[]): string => {
-    const paths: OpenAPIV3.PathsObject = {};
+  private generateOpenAPIDocument = async (metadata: ApexClassOASEligibleResponse): Promise<string> => {
+    const documentText = fs.readFileSync(new URL(metadata.resourceUri.toString()), 'utf8');
+    const className = path.basename(metadata.resourceUri, '.cls');
+    const methodNames = (metadata.symbols || [])
+      .filter((symbol: any) => symbol.isEligible)
+      .map((symbol: any) => symbol?.docSymbol?.name)
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts:99]:
``` suggestion
      .map((symbol: any) => symbol.docSymbol?.name)
```

File: packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts (hunk 3 above)
Comment (peternhale) [reply packages/salesforcedx-vscode-apex/src/commands/metadataOrchestrator.ts:62]: In addition to the selected method not being eligible, there is a chance the user did not select a method, chose gen for selected method, when they we actually trying to run for the entire class.

@CristiCanizales would you mind writing a spike work item to look into reporting of results from the eligibility call?

File: packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts [hunk 5]
```diff
@@ -39,19 +40,17 @@ export class ApexActionController {
...
 
           // Step 3: Generate OpenAPI Document
           progress.report({ message: nls.localize('generate_openapi_document') });
-          const openApiDocument = this.generateOpenAPIDocument(Array.isArray(metadata) ? metadata : [metadata]);
+          const openApiDocument = await this.generateOpenAPIDocument(metadata);
 
           // Step 4: Write OpenAPI Document to File
-          name = Array.isArray(metadata) ? metadata[0].className : metadata.name;
+          name = isClass ? path.basename(metadata.resourceUri, '.cls') : metadata?.symbols?.[0]?.docSymbol?.name;
           const openApiFileName = `${name}_openapi.yml`;
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts:54]: Name may be undefined, does undefined mean something is wrong?

File: packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts [hunk 6]
```diff
@@ -88,34 +87,17 @@ export class ApexActionController {
    * @param metadata - The metadata of the methods.
    * @returns The OpenAPI document as a string.
    */
-  private generateOpenAPIDocument = (metadata: MethodMetadata[]): string => {
-    const paths: OpenAPIV3.PathsObject = {};
+  private generateOpenAPIDocument = async (metadata: ApexClassOASEligibleResponse): Promise<string> => {
+    const documentText = fs.readFileSync(new URL(metadata.resourceUri.toString()), 'utf8');
+    const className = path.basename(metadata.resourceUri, '.cls');
+    const methodNames = (metadata.symbols || [])
+      .filter((symbol: any) => symbol.isEligible)
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts:94]: Is there a better choice than any here?

File: packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts (hunk 5 above)
Comment (CristiCanizales) [reply packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts:54]: at that point I don't think so? We'd have already generated the doc and done all the validations, so I don't know if it's that important

**Review by mingxuanzhangsfdx — APPROVED** [review]

**Pull request merged**